File: packages/core/src/forecast.ts
export interface Observation { at: number; used: number; resetAt: string | null }
import type { Forecast } from '@trackem/contracts';
export type { Forecast } from '@trackem/contracts';
/** Use only recent observations in the same quota window. No inferred start-of-window usage. */
export function forecastUsage(samples: Observation[], now = Date.now()): Forecast {
  const result: Forecast = { status: 'collecting', runsOutAt: null, sampleMinutes: 0 };
  const last = samples.at(-1);
  if (!last || !Number.isFinite(last.used)) return result;
  const reset = last.resetAt ? Date.parse(last.resetAt) : NaN;
  if (now - last.at > 15 * 60_000 || last.at > now || reset <= now) return { ...result, status: 'stale' };
  if (last.used >= 100) return { ...result, status: 'exhausted' };
  if (!Number.isFinite(reset)) return result;
  let recent = samples.filter(s => s.resetAt === last.resetAt && s.at >= now - 2 * 60 * 60_000 && s.at <= now);
  // A correction, a usage drop, or a long offline gap starts a new observation period.
  for (let i = recent.length - 1; i > 0; i--) {
    const current = recent[i];
    const previous = recent[i - 1];
    if (current && previous && (current.used < previous.used || current.at - previous.at > 15 * 60_000)) {
      recent = recent.slice(i);
      break;
    }
  }
  const first = recent[0];
  if (!first) return result;
  result.sampleMinutes = (last.at - first.at) / 60_000;
  if (recent.length < 3 || result.sampleMinutes < 10) return result;
  const delta = last.used - first.used;
  if (delta <= 0) return { ...result, status: 'idle' };
  const runsOut = last.at + (100 - last.used) / delta * (last.at - first.at);
  if (!Number.isFinite(runsOut)) return result;
  return { ...result, status: runsOut < reset ? 'depleting' : 'lasts', runsOutAt: runsOut < reset ? new Date(runsOut).toISOString() : null };
}

File: src-tauri/src/forecast.rs
use crate::types::Forecast;

#[derive(Debug, Clone)]
pub struct Observation {
    pub at: i64,
    pub used: f64,
    pub reset_at: Option<String>,
}

pub fn forecast_usage(samples: &[Observation], now: i64) -> Forecast {
    let empty = Forecast { status: "collecting".to_string(), runs_out_at: None, sample_minutes: 0.0 };
    let Some(last) = samples.last() else { return empty };
    if !last.used.is_finite() { return empty }
    let reset = last.reset_at.as_deref().and_then(parse_timestamp);
    if now - last.at > 15 * 60_000 || last.at > now || reset.is_some_and(|at| at <= now) {
        return Forecast { status: "stale".to_string(), ..empty };
    }
    if last.used >= 100.0 {
        return Forecast { status: "exhausted".to_string(), ..empty };
    }
    let Some(reset) = reset else { return empty };

    let mut recent: Vec<&Observation> = samples.iter().filter(|sample| {
        sample.reset_at == last.reset_at && sample.at >= now - 2 * 60 * 60_000 && sample.at <= now
    }).collect();
    for index in (1..recent.len()).rev() {
        let current = recent[index];
        let previous = recent[index - 1];
        if current.used < previous.used || current.at - previous.at > 15 * 60_000 {
            recent = recent.split_off(index);
            break;
        }
    }
    let Some(first) = recent.first() else { return empty };
    let sample_minutes = (last.at - first.at) as f64 / 60_000.0;
    if recent.len() < 3 || sample_minutes < 10.0 {
        return Forecast { sample_minutes, ..empty };
    }
    let delta = last.used - first.used;
    if delta <= 0.0 {
        return Forecast { status: "idle".to_string(), sample_minutes, ..empty };
    }
    let runs_out = last.at as f64 + (100.0 - last.used) / delta * (last.at - first.at) as f64;
    if !runs_out.is_finite() { return Forecast { sample_minutes, ..empty } }
    if runs_out < reset as f64 {
        Forecast { status: "depleting".to_string(), runs_out_at: Some(timestamp_from_millis(runs_out as i64)), sample_minutes }
    } else {
        Forecast { status: "lasts".to_string(), runs_out_at: None, sample_minutes }
    }
}

pub fn parse_timestamp(value: &str) -> Option<i64> {
    // RFC 3339 timestamps from provider endpoints always include a zone offset.
    let parsed = time::OffsetDateTime::parse(value, &time::format_description::well_known::Rfc3339).ok()?;
    Some(parsed.unix_timestamp_nanos() as i64 / 1_000_000)
}

pub fn timestamp_now() -> String {
    timestamp_from_millis(now_millis())
}

pub fn now_millis() -> i64 {
    std::time::SystemTime::now().duration_since(std::time::UNIX_EPOCH).map(|duration| duration.as_millis() as i64).unwrap_or_default()
}

pub fn timestamp_from_millis(value: i64) -> String {
    time::OffsetDateTime::from_unix_timestamp_nanos(value as i128 * 1_000_000)
        .map(|datetime| datetime.format(&time::format_description::well_known::Rfc3339).unwrap_or_default())
        .unwrap_or_default()
}
